### PR TITLE
Implement #231 – default slot filling

### PR DESCRIPTION
**Pull request by lemontheme**
This PR adds default slot filling as described in #231.

Here's an example (copied from the updated docs):

### Example

The template:

```htmldjango 
<div class="calendar-component">
    <div class="header">
        {% slot "header" %}Calendar header{% endslot %}
    </div>
    <div class="body">
        {% slot "body" default %}Today's date is <span>{{ date }}</span>{% endslot %}
    </div>
</div>
```

Including the component (notice how the `fill` tag is omitted):

```htmldjango
{% component_block "calendar" date="2020-06-06" %}
    Can you believe it's already <span>{{ date }}</span>??
{% endcomponent_block %}
```

The rendered result (exactly the same as before):

```html
<div class="calendar-component">
    <div class="header">
        Calendar header
    </div>
    <div class="body">
        Can you believe it's already <span>2020-06-06</span>??
    </div>
</div>
```

---

### Main changes:

- New `default` option added to `slot` tags.
- Component_block parsing updated to deal with 'bare'/'implicit' (still looking for the right term) filling.

Additional changes:

- Tests
- Updated the README + minor cosmetic changes
- Added example to sampleproject

To discuss:

- With the new `default` keyword, we're overloading the 'default' concept slightly, since it's also used in the `<fill_alias>.super` usage (i.e. where we used to have `slot.super`). In one of my code comments, I argue it might be better to drop the second use of 'default' and use `<fill_alias>.original` instead.

### CODE REVIEW

File: README.md
Comment (EmilStenstrom) [inline README.md:177]: This must be unintentional right?

File: django_components/component.py
Comment (EmilStenstrom) [inline django_components/component.py:175]: There should be a space between component and the single quote here, right?

File: django_components/component.py
Comment (EmilStenstrom) [inline django_components/component.py:182]: Too much whitespace before the comment.

File: django_components/templatetags/component_tags.py
Comment (EmilStenstrom) [inline django_components/templatetags/component_tags.py:176]: Not sure I think this is worth changing concepts for. The difference here is "default slot" vs "default content" which are different concepts. Your call, but is this worth having people change their code for? Should we change the docs?

File: django_components/templatetags/component_tags.py
Comment (EmilStenstrom) [inline django_components/templatetags/component_tags.py:229]: Strange comment, remove?

File: django_components/templatetags/component_tags.py
Comment (EmilStenstrom) [inline django_components/templatetags/component_tags.py:240]: Strange comment, remove?

File: django_components/templatetags/component_tags.py
Comment (EmilStenstrom) [inline django_components/templatetags/component_tags.py:283]: Free is not the right term here I think, maybe remove this line?

File: tests/test_templatetags.py
Comment (EmilStenstrom) [inline tests/test_templatetags.py:906]: "isvalid" -> "is valid"

File: tests/test_templatetags.py
Comment (EmilStenstrom) [inline tests/test_templatetags.py:919]: "isvalid" -> "is valid"

**Review by EmilStenstrom — APPROVED** [review]
Looks good to me! Found some small nitpicks, but nothing serious. Tests look reasonable.

If possible, it would be easier to review PR:s when unrelated changes are put in separate commits! Then I can walk through the commits one by one, and know the purpose of that commit from the commit message.

Great job, as always! :)

File: django_components/templatetags/component_tags.py
Comment (lemontheme) [reply django_components/templatetags/component_tags.py:283]: It's probably too much detail for an error msg anyway. What I mean is that 'default' can come before or after 'required'. In ebnf terms:

```ebnf
slot tag = name, [options];
options = ["default"], ["required"] | ["required"], ["default"];
```

I'll drop it and let the examples speak for themselves.

File: django_components/templatetags/component_tags.py
Comment (lemontheme) [reply django_components/templatetags/component_tags.py:176]: I think it's best not overload terms with separate meanings, even if the differences are clear to us. On the other hand, it's probably not worth breaking people's code over.

Which of these options makes most sense to you:

- The _default_ property is maintained but a new _original_ alias is introduced. (Hell, we could even go back to using _super_ =p )
- We use a different name for the slot option. I'm not sure what that would be though.
- We leave things as they are, but add a clarifiying paragraph in the docs. Maybe create a new FAQ section at the end explaining possible sources of confusion.

**Comment by lemontheme** [comment]
Thanks for reviewing. Amazing eye for detail as always!

File: django_components/templatetags/component_tags.py
Comment (lemontheme) [reply django_components/templatetags/component_tags.py:229]: Will do. (It's my way of saying 'here ends the block of code up until which the previous comment holds'.)

File: django_components/templatetags/component_tags.py
Comment (EmilStenstrom) [reply django_components/templatetags/component_tags.py:176]: I'm really not worried about this confusion. Default is a very commonly used term and being constrained to only use it in one place in the API feels limiting. It's like saying "parameter" could only be used once. But all sorts of functions take parameters.

I would say "We leave things as they are, and add things to a FAQ _IF_ it's a common confusion from users" is the best option here. But sure, it doesn't hurt to add it the FAQ if you feel strongly about it.

Comment (lemontheme) [reply django_components/templatetags/component_tags.py]: Eh, you're probably right. Until others have had the chance to use this new feature, it's too early to say what will or won't be experienced as confusing. Makes sense to hold off on updating the docs until we've received some feedback. I've dropped these lines in a subsequent commit.

**Comment by lemontheme** [comment]
Btw – don't merge this just yet. Still got a couple of tests I want to add.

**Comment by EmilStenstrom** [comment]
Rebased just to get the tests to run and report status checks again! Sorry if I made a mess.

**Comment by lemontheme** [comment]
Thanks! Was wondering why they weren't running. No mess as far as I can tell!

I think it's safe for us to merge. 

If all goes well, a rebased #273 should start reporting success as well.

**Comment by EmilStenstrom** [comment]
@lemontheme Very happy to see this merged, well done! :D Will you make a pypi release too, or wait for the other big change too?

**Comment by lemontheme** [comment]
Thanks! Took a lot of sketches and diagrams to get this right. =p

It's been a while since I've looked at the bigger picture. Am I correct in assuming the other big change you refer to is the transition to a single `component` tag (to replace `component_block`)?

If so, I believe we can do a release already, independently of that change. Essentially, what I plan to do for that other change is introduce a flag in settings that will change what function get registered to which tags. The `do_component_block()` function will be registered to `{% component %}`. `do_component()` will be ignored. `{% component_block %} will cease to exist.

Once #273 is merged succesfully, we can release.

**Pull request merged**
